packages/go_router/lib/src/builder.dart

```diff
@@ -435,6 +450,23 @@ class RouteBuilder {
                 : _errorBuilderForAppType!(context, state),
           );
   }
+
+  /// Return a HeroController based on the app type.
+  HeroController _getHeroController(BuildContext context) {
+    /// Clean up previous cache to prevent memory leak.
+    if (_goHeroCache.isNotEmpty) {
+      _goHeroCache.clear();
```

Comment:
@chunhtai Added the clean up here, as this will get called only if the NavigatorKey isn't assigned to the cache.

Reply:
Wow, it seems too aggressive cache-clearing policy 🙂. Here you clear all cache on any new element added.

I could propose the following cache-clearing policy that executes at the end of `buildPages` method:
```dart
  List<Page<Object?>> buildPages(
      BuildContext context,
      RouteMatchList matchList,
      PopPageCallback onPopPage,
      bool routerNeglect,
      GlobalKey<NavigatorState> navigatorKey,
      Map<Page<Object?>, GoRouterState> registry) {
    final Map<GlobalKey<NavigatorState>, List<Page<Object?>>> keyToPage =
        <GlobalKey<NavigatorState>, List<Page<Object?>>>{};
    try {
      _buildRecursive(context, matchList, 0, onPopPage, routerNeglect,
          keyToPage, navigatorKey, registry);
      return keyToPage[navigatorKey]!;
    } on _RouteBuilderError catch (e) {
      return <Page<Object?>>[
        _buildErrorPage(context, e, matchList.uri),
      ];
    } finally {
      _goHeroCache.removeWhere((key, _) => !keyToPage.keys.contains(key));
    }
  }
```

Reply:
Thanks, added your proposed cache-policy.